connector/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java

```diff
@@ -0,0 +1,416 @@
+  // users,
+  // but if they become available, it would be advisable to append an annotation to the
+  // protoFieldBuilder
+  // for these and other types.
+  private static DescriptorProtos.FieldDescriptorProto.Type toProtoFieldType(
+      LegacySQLTypeName bqType) {
+    DescriptorProtos.FieldDescriptorProto.Type protoFieldType;
+    if (LegacySQLTypeName.INTEGER.equals(bqType)
+        || LegacySQLTypeName.DATE.equals(bqType)
+        || LegacySQLTypeName.DATETIME.equals(bqType)
+        || LegacySQLTypeName.TIMESTAMP.equals(bqType)) {
+      return DescriptorProtos.FieldDescriptorProto.Type.TYPE_INT64;
+    }
+    if (LegacySQLTypeName.BOOLEAN.equals(bqType)) {
+      return DescriptorProtos.FieldDescriptorProto.Type.TYPE_BOOL;
+    }
+    if (LegacySQLTypeName.STRING.equals(bqType)) {
+      return DescriptorProtos.FieldDescriptorProto.Type.TYPE_STRING;
+    }
+    if (LegacySQLTypeName.GEOGRAPHY.equals(bqType)
+        || LegacySQLTypeName.BYTES.equals(bqType)
+        || LegacySQLTypeName.NUMERIC.equals(bqType)) {
+      return DescriptorProtos.FieldDescriptorProto.Type.TYPE_BYTES;
+    }
+    if (LegacySQLTypeName.FLOAT.equals(bqType)) {
+      return DescriptorProtos.FieldDescriptorProto.Type.TYPE_DOUBLE;
+    }
+
+    if (LegacySQLTypeName.RECORD.equals(bqType)) {
+      throw new IllegalStateException(
+          "Program attempted to return an atomic data-type for a RECORD");
+    }
+
+    throw new IllegalArgumentException("Unexpected type: " + bqType.name());
+  }
+
+  /**
+   * Spark Row --> ProtoRows converter utils: To be used by the DataWriters facing the BigQuery
+   * Storage Write API
+   */
+  public static ProtoBufProto.ProtoRows toProtoRows(StructType sparkSchema, InternalRow[] rows) {
+    try {
+      Descriptors.Descriptor schemaDescriptor = toDescriptor(sparkSchema);
+      ProtoBufProto.ProtoRows.Builder protoRows = ProtoBufProto.ProtoRows.newBuilder();
+      for (InternalRow row : rows) {
+        DynamicMessage rowMessage = buildSingleRowMessage(sparkSchema, schemaDescriptor, row);
+        protoRows.addSerializedRows(rowMessage.toByteString());
+      }
+      return protoRows.build();
+    } catch (Exception e) {
+      throw new RuntimeException("Could not convert Internal Rows to Proto Rows.", e);
+    }
+  }
+
+  public static DynamicMessage buildSingleRowMessage(
+      StructType schema, Descriptors.Descriptor schemaDescriptor, InternalRow row) {
+    DynamicMessage.Builder messageBuilder = DynamicMessage.newBuilder(schemaDescriptor);
+
+    for (int fieldIndex = 0; fieldIndex < schemaDescriptor.getFields().size(); fieldIndex++) {
+      StructField sparkField = schema.fields()[fieldIndex];
+      DataType sparkType = sparkField.dataType();
+
+      Object sparkValue = row.get(fieldIndex, sparkType);
+      boolean nullable = sparkField.nullable();
+      Descriptors.Descriptor nestedTypeDescriptor =
+          schemaDescriptor.findNestedTypeByName(RESERVED_NESTED_TYPE_NAME + (fieldIndex + 1));
```

Comment:
please create a local variable `int protoFieldNumber = fieldIndex + 1` and use it instead of the explicit addition. It makes the code more readable.